torch/csrc/jit/codegen/cuda/lower_predicate_elimination.cpp

```diff
@@ -193,6 +193,20 @@ class PredicateChcker : public IterVisitor {
         predicateNonDivisibleRootDomains(expr) ||
         predicateNonDivisibleSplit(expr);
 
+    // A cp.async op would need a predicate for either the global
+    //  input or its shared mem output, or both.
+    // Due to the WAR discussed in [Predicate Inversion for CpAsync],
+    //  we currently cannot support use cases where both the gmem read
+    //  and the smem write need to be predicated.
```

Comment:
This just requires a predicate for the GMEM read no? I'm failing to see how this is different from the logic/limitation in my comment in `lower_predicate.cpp`. We would want to read valid GMEM locations which would have to have a valid associated SMEM write location. So the "valid region" is the same for GMEM and SMEM and guaranteed to be in "allocated memory". I think this is the same limitation, just wasn't immediately obvious to me.

Reply:
Yes this is the same limitation and the assertion is put here because here we have access to the `allocated region` info. If we have `valid data` we are guaranteed to have `allocated region` but due to the limitation discussed in `lower_predicate.cpp`, for cp.async we just also require `allocated region` for all active threads in this instruction. 

I will still need to work through pathological cases in follow ups, for functional coverage but not perf critical use case.